From c913d0889c0538f7c64fc9fc64580c5d1e9eeeb7 Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Wed, 26 Mar 2025 16:33:22 -0400
Subject: [PATCH] Issue 6695 - UI - fix more minor issues

Description:

fix the following items:

MemberOf page - subtree scopes - allows you to create blank values, and does not validate suffixes
Local password policy - copy/paste error added a duplicate label for a checkbox (send expiring warning)
Create instance modal - did not validate database name
Update connection monitor chart to include detailed information for connection status (established, close wait, and time wait)

relates: https://github.com/389ds/389-ds-base/issues/6695

Reviewed by: spichugi(Thanks!)
---
 src/cockpit/389-console/src/dsModals.jsx      |  22 ++--
 .../389-console/src/lib/database/localPwp.jsx |   3 -
 .../src/lib/monitor/serverMonitor.jsx         | 114 +++++++++++++++---
 .../389-console/src/lib/plugins/memberOf.jsx  |  18 +--
 src/lib389/lib389/monitor.py                  |  12 ++
 5 files changed, 128 insertions(+), 41 deletions(-)

diff --git a/src/cockpit/389-console/src/dsModals.jsx b/src/cockpit/389-console/src/dsModals.jsx
index e3ee18fc0..bf19ce144 100644
--- a/src/cockpit/389-console/src/dsModals.jsx
+++ b/src/cockpit/389-console/src/dsModals.jsx
@@ -4,7 +4,13 @@ import PropTypes from "prop-types";
 import { DoubleConfirmModal } from "./lib/notifications.jsx";
 import { BackupTable } from "./lib/database/databaseTables.jsx";
 import { BackupModal } from "./lib/database/backups.jsx";
-import { log_cmd, bad_file_name, valid_dn, callCmdStreamPassword } from "./lib/tools.jsx";
+import {
+    log_cmd,
+    bad_file_name,
+    valid_dn,
+    valid_db_name,
+    callCmdStreamPassword
+} from "./lib/tools.jsx";
 import {
     Button,
     Checkbox,
@@ -103,10 +109,6 @@ export class CreateInstanceModal extends React.Component {
             'createDM'
         ];
 
-        const optionalAttrs = [
-            'createDBName'
-        ];
-
         // Handle server ID
         if (this.state.createServerId !== "") {
             if (this.state.createServerId.length > 80) {
@@ -142,11 +144,9 @@ export class CreateInstanceModal extends React.Component {
         }
 
         if (this.state.createDBCheckbox) {
-            for (const attr of optionalAttrs) {
-                if (this.state[attr] === "") {
-                    all_good = false;
-                    errObj[attr] = true;
-                }
+            if (!valid_db_name(this.state.createDBName)) {
+                all_good = false;
+                errObj["createDBName"] = true;
             }
             if (!valid_dn(this.state.createDBSuffix)) {
                 all_good = false;
@@ -636,7 +636,7 @@ export class CreateInstanceModal extends React.Component {
                                         <FormHelperText >
                                             <HelperText>
                                                 <HelperTextItem variant="error">
-                                                    {_("Name is required")}
+                                                    {createDBName === "" ? _("Name is required") : "Invalid database name"}
                                                 </HelperTextItem>
                                             </HelperText>
                                         </FormHelperText>
diff --git a/src/cockpit/389-console/src/lib/database/localPwp.jsx b/src/cockpit/389-console/src/lib/database/localPwp.jsx
index ff114bb01..8586ba932 100644
--- a/src/cockpit/389-console/src/lib/database/localPwp.jsx
+++ b/src/cockpit/389-console/src/lib/database/localPwp.jsx
@@ -398,9 +398,6 @@ class CreatePolicy extends React.Component {
                                     </GridItem>
                                 </Grid>
                                 <Grid className="ds-margin-top" title={_("Always return a password expiring control when requested (passwordSendExpiringTime).")}>
-                                    <GridItem className="ds-label" span={4}>
-                                        {_("Send Password Expiring Warning")}
-                                    </GridItem>
                                     <GridItem span={4}>
                                         <Checkbox
                                             id="create_passwordsendexpiringtime"
diff --git a/src/cockpit/389-console/src/lib/monitor/serverMonitor.jsx b/src/cockpit/389-console/src/lib/monitor/serverMonitor.jsx
index 513e1d792..a736ff532 100644
--- a/src/cockpit/389-console/src/lib/monitor/serverMonitor.jsx
+++ b/src/cockpit/389-console/src/lib/monitor/serverMonitor.jsx
@@ -37,6 +37,7 @@ import {
 import { SyncAltIcon } from "@patternfly/react-icons";
 
 const _ = cockpit.gettext;
+const refresh_interval = 10000; // 10 seconds
 
 export class ServerMonitor extends React.Component {
     constructor (props) {
@@ -47,12 +48,20 @@ export class ServerMonitor extends React.Component {
         const initResChart = [];
         const initSwapChart = [];
         const initConnChart = [];
-        for (let idx = 1; idx <= 6; idx++) {
-            initCPUChart.push({ name: 'CPU', x: "0:00:00", y: 0 });
-            initResChart.push({ name: 'Resident', x: "0:00:00", y: 0 });
-            initVirtChart.push({ name: 'Virtual', x: "0:00:00", y: 0 });
-            initSwapChart.push({ name: 'Swap', x: "0:00:00", y: 0 });
-            initConnChart.push({ name: 'Connection', x: "0:00:00", y: 0 });
+        const initConnEstablishedChart = [];
+        const initConnTimeWaitChart = [];
+        const initConnCloseWaitChart = [];
+        for (let idx = 0; idx <= 5; idx++) {
+            const value = refresh_interval / 1000;
+            const x_value = "0:00:" + (idx === 0 ? "00" : value * idx).toString();
+            initCPUChart.push({ name: 'CPU', x: x_value, y: 0 });
+            initResChart.push({ name: 'Resident', x: x_value, y: 0 });
+            initVirtChart.push({ name: 'Virtual', x: x_value, y: 0 });
+            initSwapChart.push({ name: 'Swap', x: x_value, y: 0 });
+            initConnChart.push({ name: 'Connections', x: x_value, y: 0 });
+            initConnTimeWaitChart.push({ name: 'Connections time wait', x: x_value, y: 0 });
+            initConnCloseWaitChart.push({ name: 'Connections close wait', x: x_value, y: 0 });
+            initConnEstablishedChart.push({ name: 'Connections established', x: x_value, y: 0 });
         }
 
         this.state = {
@@ -71,11 +80,17 @@ export class ServerMonitor extends React.Component {
             initResChart,
             initSwapChart,
             initConnChart,
+            initConnEstablishedChart,
+            initConnTimeWaitChart,
+            initConnCloseWaitChart,
             cpuChart: [...initCPUChart],
             memVirtChart: [...initVirtChart],
             memResChart: [...initResChart],
             swapChart: [...initSwapChart],
             connChart: [...initConnChart],
+            connEstablishedChart: [...initConnEstablishedChart],
+            connTimeWaitChart: [...initConnTimeWaitChart],
+            connCloseWaitChart: [...initConnCloseWaitChart],
         };
 
         this.handleNavSelect = (event, tabIndex) => {
@@ -110,6 +125,9 @@ export class ServerMonitor extends React.Component {
             memResChart: [...this.state.initResChart],
             swapChart: [...this.state.initSwapChart],
             connChart: [...this.state.initConnChart],
+            connCloseWaitChart: [...this.state.initConnCloseWaitChart],
+            connTimeWaitChart: [...this.state.initConnTimeWaitChart],
+            connEstablishedChart: [...this.state.initConnEstablishedChart],
         });
     }
 
@@ -123,6 +141,9 @@ export class ServerMonitor extends React.Component {
         let res_mem = 0;
         let swap_mem = 0;
         let current_conns = 0;
+        let conn_established = 0;
+        let conn_close_wait = 0;
+        let conn_time_wait = 0;
         let total_threads = 0;
         let conn_highmark = this.state.conn_highmark;
         let cpu_tick_values = this.state.cpu_tick_values;
@@ -147,6 +168,9 @@ export class ServerMonitor extends React.Component {
                     res_mem = attrs['rss'][0];
                     swap_mem = attrs['swap'][0];
                     current_conns = attrs['connection_count'][0];
+                    conn_established = attrs['connection_established_count'][0];
+                    conn_close_wait = attrs['connection_close_wait_count'][0];
+                    conn_time_wait = attrs['connection_time_wait_count'][0];
                     total_threads = attrs['total_threads'][0];
                     mem_total = attrs['total_mem'][0];
 
@@ -196,6 +220,18 @@ export class ServerMonitor extends React.Component {
                     connChart.shift();
                     connChart.push({ name: _("Connections"), x: interval, y: parseInt(current_conns) });
 
+                    const connEstablishedChart = this.state.connEstablishedChart;
+                    connEstablishedChart.shift();
+                    connEstablishedChart.push({ name: _("Connections established"), x: interval, y: parseInt(conn_established) });
+
+                    const connTimeWaitChart = this.state.connTimeWaitChart;
+                    connTimeWaitChart.shift();
+                    connTimeWaitChart.push({ name: _("Connections time wait"), x: interval, y: parseInt(conn_time_wait) });
+
+                    const connCloseWaitChart = this.state.connCloseWaitChart;
+                    connCloseWaitChart.shift();
+                    connCloseWaitChart.push({ name: _("Connections close wait"), x: interval, y: parseInt(conn_close_wait) });
+
                     this.setState({
                         cpu_tick_values,
                         conn_tick_values,
@@ -204,8 +240,14 @@ export class ServerMonitor extends React.Component {
                         memResChart,
                         swapChart,
                         connChart,
+                        connTimeWaitChart,
+                        connCloseWaitChart,
+                        connEstablishedChart,
                         conn_highmark,
                         current_conns,
+                        conn_close_wait,
+                        conn_time_wait,
+                        conn_established,
                         mem_virt_size: virt_mem,
                         mem_res_size: res_mem,
                         mem_swap_size: swap_mem,
@@ -224,7 +266,7 @@ export class ServerMonitor extends React.Component {
 
     startRefresh() {
         this.setState({
-            chart_refresh: setInterval(this.refreshCharts, 10000),
+            chart_refresh: setInterval(this.refreshCharts, refresh_interval),
         });
     }
 
@@ -236,8 +278,14 @@ export class ServerMonitor extends React.Component {
         const {
             cpu,
             connChart,
+            connTimeWaitChart,
+            connCloseWaitChart,
+            connEstablishedChart,
             cpuChart,
             current_conns,
+            conn_established,
+            conn_close_wait,
+            conn_time_wait,
             memResChart,
             memVirtChart,
             swapChart,
@@ -312,15 +360,33 @@ export class ServerMonitor extends React.Component {
                                 <Card className="ds-margin-top-lg">
                                     <CardBody>
                                         <Grid>
-                                            <GridItem span="4" className="ds-center" title={_("Established client connections to the server")}>
-                                                <TextContent>
-                                                    <Text className="ds-margin-top-xlg" component={TextVariants.h3}>
-                                                        {_("Connections")}
+                                            <GridItem span="4" title={_("Established client connections to the server")}>
+                                                <div className="ds-center" >
+                                                    <TextContent>
+                                                        <Text component={TextVariants.h2}>
+                                                            {_("Connections")}
+                                                        </Text>
+                                                    </TextContent>
+                                                    <TextContent>
+                                                        <Text component={TextVariants.h6}>
+                                                            <b>{numToCommas(current_conns)}</b>
+                                                        </Text>
+                                                    </TextContent>
+                                                    <Divider className="ds-margin-top ds-margin-bottom"/>
+                                                </div>
+                                                <TextContent className="ds-margin-top-lg" title="Connections that are in an ESTABLISHED state">
+                                                    <Text component={TextVariants.p}>
+                                                        Established: &nbsp;&nbsp;<b>{numToCommas(conn_established)}</b>
                                                     </Text>
                                                 </TextContent>
-                                                <TextContent>
-                                                    <Text component={TextVariants.h6}>
-                                                        <b>{numToCommas(current_conns)}</b>
+                                                <TextContent className="ds-margin-top-lg" title="Connections that are in a CLOSE_WAIT state">
+                                                    <Text component={TextVariants.p}>
+                                                        Close wait: &nbsp;&nbsp;<b>{numToCommas(conn_close_wait)}</b>
+                                                    </Text>
+                                                </TextContent>
+                                                <TextContent className="ds-margin-top-lg" title="Connections that are in a TIME_WAIT state">
+                                                    <Text component={TextVariants.p}>
+                                                        Time wait: &nbsp;&nbsp;<b>{numToCommas(conn_time_wait)}</b>
                                                     </Text>
                                                 </TextContent>
                                             </GridItem>
@@ -329,13 +395,13 @@ export class ServerMonitor extends React.Component {
                                                     ariaDesc="connection stats"
                                                     ariaTitle={_("Live Connection Statistics")}
                                                     containerComponent={<ChartVoronoiContainer labels={({ datum }) => `${datum.name}: ${datum.y}`} constrainToVisibleArea />}
-                                                    height={200}
+                                                    height={220}
                                                     minDomain={{ y: 0 }}
                                                     padding={{
                                                         bottom: 30,
-                                                        left: 55,
+                                                        left: 60,
                                                         top: 10,
-                                                        right: 25,
+                                                        right: 30,
                                                     }}
                                                 >
                                                     <ChartAxis />
@@ -344,6 +410,18 @@ export class ServerMonitor extends React.Component {
                                                         <ChartArea
                                                             data={connChart}
                                                         />
+                                                        <ChartArea
+                                                            data={connEstablishedChart}
+                                                            interpolation="monotoneX"
+                                                        />
+                                                        <ChartArea
+                                                            data={connTimeWaitChart}
+                                                            interpolation="monotoneX"
+                                                        />
+                                                        <ChartArea
+                                                            data={connCloseWaitChart}
+                                                            interpolation="monotoneX"
+                                                        />
                                                     </ChartGroup>
                                                 </Chart>
                                             </GridItem>
@@ -372,7 +450,7 @@ export class ServerMonitor extends React.Component {
                                                     ariaDesc="cpu"
                                                     ariaTitle={_("Server CPU Usage")}
                                                     containerComponent={<ChartVoronoiContainer labels={({ datum }) => `${datum.name}: ${datum.y}%`} constrainToVisibleArea />}
-                                                    height={200}
+                                                    height={220}
                                                     minDomain={{ y: 0 }}
                                                     padding={{
                                                         bottom: 30,
diff --git a/src/cockpit/389-console/src/lib/plugins/memberOf.jsx b/src/cockpit/389-console/src/lib/plugins/memberOf.jsx
index 7c1b02297..704d6d0b1 100644
--- a/src/cockpit/389-console/src/lib/plugins/memberOf.jsx
+++ b/src/cockpit/389-console/src/lib/plugins/memberOf.jsx
@@ -225,7 +225,7 @@ class MemberOf extends React.Component {
 
         // Handle scope subtree
         this.handleSubtreeScopeSelect = (event, selection) => {
-            if (selection === "") {
+            if (!selection.trim() || !valid_dn(selection)) {
                 this.setState({isSubtreeScopeOpen: false});
                 return;
             }
@@ -257,7 +257,7 @@ class MemberOf extends React.Component {
             }, () => { this.validateConfig() });
         };
         this.handleSubtreeScopeCreateOption = newValue => {
-            if (newValue && !this.state.memberOfEntryScopeOptions.includes(newValue)) {
+            if (newValue.trim() && valid_dn(newValue) && !this.state.memberOfEntryScopeOptions.includes(newValue)) {
                 this.setState({
                     memberOfEntryScopeOptions: [...this.state.memberOfEntryScopeOptions, newValue],
                     isSubtreeScopeOpen: false
@@ -267,7 +267,7 @@ class MemberOf extends React.Component {
 
         // Handle Exclude Scope subtree
         this.handleExcludeScopeSelect = (event, selection) => {
-            if (selection === "") {
+            if (!selection.trim() || !valid_dn(selection)) {
                 this.setState({isExcludeScopeOpen: false});
                 return;
             }
@@ -299,7 +299,7 @@ class MemberOf extends React.Component {
             }, () => { this.validateConfig() });
         };
         this.handleExcludeCreateOption = newValue => {
-            if (newValue && !this.state.memberOfEntryScopeOptions.includes(newValue)) {
+            if (newValue.trim() && valid_dn(newValue) && !this.state.memberOfEntryScopeOptions.includes(newValue)) {
                 this.setState({
                     memberOfEntryScopeExcludeOptions: [...this.state.memberOfEntryScopeExcludeOptions, newValue],
                     isExcludeScopeOpen: false
@@ -310,7 +310,7 @@ class MemberOf extends React.Component {
         // Modal scope and exclude Scope
         // Handle scope subtree
         this.handleConfigScopeSelect = (event, selection) => {
-            if (selection === "") {
+            if (selection.trim() === "" || !valid_dn(selection)) {
                 this.setState({isConfigSubtreeScopeOpen: false});
                 return;
             }
@@ -342,7 +342,7 @@ class MemberOf extends React.Component {
             }, () => { this.validateModal() });
         };
         this.handleConfigCreateOption = newValue => {
-            if (newValue && !this.state.configEntryScopeOptions.includes(newValue)) {
+            if (newValue.trim() && valid_dn(newValue) && !this.state.configEntryScopeOptions.includes(newValue)) {
                 this.setState({
                     configEntryScopeOptions: [...this.state.configEntryScopeOptions, newValue],
                     isConfigSubtreeScopeOpen: false
@@ -352,7 +352,7 @@ class MemberOf extends React.Component {
 
         // Handle Exclude Scope subtree
         this.handleConfigExcludeScopeSelect = (event, selection) => {
-            if (selection === "") {
+            if (selection.trim() === "" || !valid_dn(selection)) {
                 this.setState({isConfigExcludeScopeOpen: false});
                 return;
             }
@@ -384,7 +384,7 @@ class MemberOf extends React.Component {
             }, () => { this.validateModal() });
         };
         this.handleConfigExcludeCreateOption = newValue => {
-            if (newValue && !this.state.configEntryScopeExcludeOptions.includes(newValue)) {
+            if (newValue.trim() && valid_dn(newValue) &&  !this.state.configEntryScopeExcludeOptions.includes(newValue)) {
                 this.setState({
                     configEntryScopeExcludeOptions: [...this.state.configEntryScopeExcludeOptions, newValue],
                     isConfigExcludeScopeOpen: false
@@ -1563,7 +1563,7 @@ class MemberOf extends React.Component {
                                     ))}
                                 </Select>
                                 <FormHelperText  >
-                                    {_("A subtree is required, and values must be valid DN's")}
+                                    {"values must be valid DN's"}
                                 </FormHelperText>
                             </GridItem>
                             <GridItem className="ds-left-margin" span={3}>
diff --git a/src/lib389/lib389/monitor.py b/src/lib389/lib389/monitor.py
index 196577ed5..ec82b0346 100644
--- a/src/lib389/lib389/monitor.py
+++ b/src/lib389/lib389/monitor.py
@@ -119,14 +119,26 @@ class Monitor(DSLdapObject):
             sslport = str(self._instance.sslport)
 
         conn_count = 0
+        conn_established_count = 0
+        conn_close_wait_count = 0
+        conn_time_wait_count = 0
         conns = psutil.net_connections()
         for conn in conns:
             if len(conn[4]) > 0:
                 conn_port = str(conn[4][1])
                 if conn_port in (port, sslport):
+                    if conn[5] == 'TIME_WAIT':
+                        conn_time_wait_count += 1
+                    if conn[5] == 'CLOSE_WAIT':
+                        conn_close_wait_count += 1
+                    if conn[5] == 'ESTABLISHED':
+                        conn_established_count += 1
                     conn_count += 1
 
         stats['connection_count'] = [str(conn_count)]
+        stats['connection_established_count'] = [str(conn_established_count)]
+        stats['connection_close_wait_count'] = [str(conn_close_wait_count)]
+        stats['connection_time_wait_count'] = [str(conn_time_wait_count)]
 
         return stats
 
-- 
2.48.1

